### ISSUE TITLE
Wan iteration time and loss curve

### ISSUE THREAD

**Issue by SagiPolaczek** [tdrussell/diffusion-pipe#410]
Hey!
Firstly, thank you for creating and maintaining this package!

I'm training Wan2.1 on a custom **video** dataset - each video is 80x960x480 ~ and I'm using 4xA6000 node for a single training session.
I have two questions please:

1. Does the loss curve should "converge nicely"? or is it expected to look noisy? Do you have any example to share?
2. I noticed that the iteration time is very high. Are my numbers normal? Am I missing something? I'm considering to try small amounts of frames per sample (maybe several buckets?).

Thanks in advance! Hopefully this discussion will help others as well.

```
[2025-09-04 15:18:01,200] [INFO] [logging.py:107:log_dist] [Rank 0] step=1, skipped=0, lr=[1.5968e-07], mom=[0.0]
steps: 1 loss: 0.0614 iter time (s): 531.283 samples/sec: 0.008
[2025-09-04 15:26:58,142] [INFO] [logging.py:107:log_dist] [Rank 0] step=2, skipped=0, lr=[2.3936e-07], mom=[0.0]
steps: 2 loss: 0.0571 iter time (s): 532.811 samples/sec: 0.008
[2025-09-04 15:35:55,125] [INFO] [logging.py:107:log_dist] [Rank 0] step=3, skipped=0, lr=[3.1904000000000003e-07], mom=[0.0]
steps: 3 loss: 0.0359 iter time (s): 533.063 samples/sec: 0.008
[2025-09-04 15:44:52,818] [INFO] [logging.py:107:log_dist] [Rank 0] step=4, skipped=0, lr=[3.9872000000000007e-07], mom=[0.0]
steps: 4 loss: 0.1193 iter time (s): 533.263 samples/sec: 0.008
[2025-09-04 15:53:51,236] [INFO] [logging.py:107:log_dist] [Rank 0] step=5, skipped=0, lr=[4.784000000000002e-07], mom=[0.0]
steps: 5 loss: 0.0492 iter time (s): 533.754 samples/sec: 0.007
[2025-09-04 16:02:49,414] [INFO] [logging.py:107:log_dist] [Rank 0] step=6, skipped=0, lr=[5.580800000000002e-07], mom=[0.0]
steps: 6 loss: 0.1036 iter time (s): 533.689 samples/sec: 0.007
[2025-09-04 16:11:45,821] [INFO] [logging.py:107:log_dist] [Rank 0] step=7, skipped=0, lr=[6.377600000000002e-07], mom=[0.0]
steps: 7 loss: 0.0882 iter time (s): 533.204 samples/sec: 0.008
[2025-09-04 16:20:42,221] [INFO] [logging.py:107:log_dist] [Rank 0] step=8, skipped=0, lr=[7.174400000000003e-07], mom=[0.0]
steps: 8 loss: 0.0402 iter time (s): 532.987 samples/sec: 0.008
[2025-09-04 16:29:39,502] [INFO] [logging.py:107:log_dist] [Rank 0] step=9, skipped=0, lr=[7.971200000000003e-07], mom=[0.0]
steps: 9 loss: 0.0551 iter time (s): 533.313 samples/sec: 0.008
[2025-09-04 16:38:35,661] [INFO] [logging.py:107:log_dist] [Rank 0] step=10, skipped=0, lr=[8.768000000000003e-07], mom=[0.0]
steps: 10 loss: 0.0570 iter time (s): 532.878 samples/sec: 0.008
[2025-09-04 16:47:32,394] [INFO] [logging.py:107:log_dist] [Rank 0] step=11, skipped=0, lr=[9.564800000000004e-07], mom=[0.0]
steps: 11 loss: 0.0957 iter time (s): 532.916 samples/sec: 0.008
[2025-09-04 16:56:29,250] [INFO] [logging.py:107:log_dist] [Rank 0] step=12, skipped=0, lr=[1.0361600000000005e-06], mom=[0.0]
steps: 12 loss: 0.0445 iter time (s): 533.085 samples/sec: 0.008
[2025-09-04 17:05:26,164] [INFO] [logging.py:107:log_dist] [Rank 0] step=13, skipped=0, lr=[1.1158400000000004e-06], mom=[0.0]
steps: 13 loss: 0.0746 iter time (s): 533.286 samples/sec: 0.008
```

**Comment by Oruli** [comment]
You need to look at epoch_loss not per step. Use tensorboard.

**Comment by SagiPolaczek** [comment]
@Oruli Thanks for the comment!

So yeah I've been looking at the epoch loss. Still not sure what a good trend should look like (if any, for fine-tuning).

For example:
<img width="1328" height="432" alt="Image" src="https://github.com/user-attachments/assets/e2a66edc-a491-4b54-b69d-716ea66f7854" />
While the epoch doesn't converge the evaluation results do improve.



Also, I will take this message as an opportunity to update regarding the training time:
Turns out that the GPU topology of the node I've been using was poor. The connections between the GPU themselves wasn't fast -> causing a huge overhead in every step.
I moved to a different node, slightly better, and got better timings:
```
[2025-09-09 11:43:38,808] [INFO] [logging.py:107:log_dist] [Rank 0] step=1095, skipped=0, lr=[2e-05], mom=[0.0]
2025-09-09 11:43:38
steps: 1095 loss: 0.0368 iter time (s): 114.383 samples/sec: 0.035
2025-09-09 11:46:22
[2025-09-09 11:46:22,737] [INFO] [logging.py:107:log_dist] [Rank 0] step=1096, skipped=0, lr=[2e-05], mom=[0.0]
2025-09-09 11:46:22
steps: 1096 loss: 0.0791 iter time (s): 158.851 samples/sec: 0.025
2025-09-09 11:49:05
[2025-09-09 11:49:05,595] [INFO] [logging.py:107:log_dist] [Rank 0] step=1097, skipped=0, lr=[2e-05], mom=[0.0]
2025-09-09 11:49:05
steps: 1097 loss: 0.0962 iter time (s): 158.838 samples/sec: 0.025
```

Probably still sub-optimal, but that's good enough for now.

**Comment by Oruli** [comment]
@SagiPolaczek I'm no pro by any stretch, also just learning this element myself. Your eval chart should be more telling, you want to find the lowest part just before it starts going up. Does the chart look good? It should be a nice curve sharp at the beginning an then taper out

**Comment by SagiPolaczek** [comment]
That's cool, we are all constantly learning together.

I don't use eval set since it's too costly. Instead, I'm sampling videos from the checkpoints I've created. The cool part is that I'm doing it on a different machine - so I don't waste training time. (will try to submit a PR soon)

So in my case, the training loss look kind of the same - but the video that I'm generating (both training and evaluation) are constantly improving.

**Comment by Oruli** [comment]
> That's cool, we are all constantly learning together.
> 
> I don't use eval set since it's too costly. Instead, I'm sampling videos from the checkpoints I've created. The cool part is that I'm doing it on a different machine - so I don't waste training time. (will try to submit a PR soon)
> 
> So in my case, the training loss look kind of the same - but the video that I'm generating (both training and evaluation) are constantly improving.

You need the eval set, unfortunately. No other metric will give you what you need. I appreciate what you say about being costly, but you can have a very small set and only run it every 5 epoch or whenever you want to keep the time down. For me on a 150 image dataset it I have eval at 20 images and it only takes about 30-40 seconds with my batch settings

See the output below, this is what you want. 1200 steps roughly where it's cooked

<img width="1737" height="383" alt="Image" src="https://github.com/user-attachments/assets/5c37a632-112c-4f95-b274-28cccb8f3073" />

**Comment by SagiPolaczek** [comment]
I work with videos, using Wan2.1 14B I2V model so:

1. I wonder if such graph is what I'm expecting to get there. I'm not sure..
2. The timings much differ. 

I will give it a try again tho. Thanks a lot!

**Comment by tedbiv** [comment]
i think this is from a wan2.1 lora. 1024x1024x90frame at 16fps videos down sampled to 128x128 due to vram constraint. but this is my normal training curve. 

<img width="2954" height="1864" alt="Image" src="https://github.com/user-attachments/assets/febc1943-fedf-44b7-b6de-ed6f6c024868" />

**Comment by alienzed** [comment]
WAN2.1 is noisy... not all of my lora had curves that went clearly downward. Some did though. I started training with HunyuanVideo and the curves were unmistakably downward, unless there was a major issue. For WAN, the first 5 I did, the curve looked darn flat.

You can determine good epochs from loss curve alone, but it's tough. You find the various floors that segments oscillate around, on those floors, you look for valleys (lower area) and in those valleys you look for stability, meaning low-ish loss where the siblings are not spikes or trenches. Sometimes a magic epoch at very low loss will appear, but I've found some of my best performing epochs were a touch above the lowest losses in a valley at a floor. The eval set, which I've yet to try (trying now actually!) is supposed to tell you where that valley/floor combo is, so that you can test the stable low loss epochs inside it.
I'll add that 960x480x80 is huge, really massive. Sure, you're throwing enough VRAM at it, but imo that's overkill. I only have a 5090 9now, wooo), but I train with half the frames and usually below that resolution and my LORAs are quite good :) that's subjective though...
